lib/auth/mfa/mfav1/service.go

```diff
@@ -16,18 +16,430 @@
+	if payload == nil {
+		return trace.BadParameter("missing CreateSessionChallengeRequest payload")
+	}
+
+	if len(payload.GetSshSessionId()) == 0 {
+		return trace.BadParameter("empty SshSessionId in payload")
+	}
+
+	// If either SSO challenge field is set, both must be set.
+	if req.SsoClientRedirectUrl != "" || req.ProxyAddressForSso != "" {
+		if req.SsoClientRedirectUrl == "" {
+			return trace.BadParameter("missing SsoClientRedirectUrl for SSO challenge")
+		}
+
+		if req.ProxyAddressForSso == "" {
+			return trace.BadParameter("missing ProxyAddressForSso for SSO challenge")
+		}
+	}
+
+	return nil
+}
+
+func groupByDeviceType(devices []*types.MFADevice) ([]*types.MFADevice, *types.MFADevice) {
+	var (
+		webauthnDevices []*types.MFADevice
+		ssoDevice       *types.MFADevice
+	)
+
+	// Skip unsupported device types. For example, TOTP devices are not supported for session-based MFA challenges and
+	// should be ignored.
+	for _, dev := range devices {
+		switch dev.Device.(type) {
+		case *types.MFADevice_U2F, *types.MFADevice_Webauthn:
+			webauthnDevices = append(webauthnDevices, dev)
+		case *types.MFADevice_Sso:
+			ssoDevice = dev
+		}
+	}
+
+	return webauthnDevices, ssoDevice
+}
+
+func validateValidateSessionChallengeRequest(req *mfav1.ValidateSessionChallengeRequest) error {
+	if req == nil {
+		return trace.BadParameter("param ValidateSessionChallengeRequest is nil")
+	}
+
+	mfaResp := req.GetMfaResponse()
+	if mfaResp == nil {
+		return trace.BadParameter("missing ValidateSessionChallengeRequest response")
+	}
+
+	resp := mfaResp.GetResponse()
+	if resp == nil {
+		return trace.BadParameter("missing MFAResponse response")
+	}
+
+	if mfaResp.GetWebauthn() == nil && mfaResp.GetSso() == nil {
+		return trace.BadParameter("at least one of WebauthnResponse or SSOResponse must be provided")
+	}
+
+	return nil
+}
+
+func (s *Service) validateWebauthnResponse(
+	ctx context.Context,
+	username string,
+	resp *mfav1.AuthenticateResponse_Webauthn,
+) (*types.MFADevice, error) {
+	pref, err := s.cache.GetAuthPreference(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Get the user's U2F preference. If it doesn't exist, continue since U2F may not be enabled.
+	u2fPref, err := pref.GetU2F()
+	if err != nil && !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+
+	// Get the user's Webauthn preference. If it doesn't exist, continue since Webauthn may not be enabled.
+	webConfig, err := pref.GetWebauthn()
+	if err != nil && !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+
+	webLogin := &wanlib.LoginFlow{
+		U2F:      u2fPref,
+		Webauthn: webConfig,
+		Identity: s.identity,
+	}
+
+	loginData, err := webLogin.Finish(
+		ctx,
+		username,
+		wantypes.CredentialAssertionResponseFromProto(resp.Webauthn),
+		&mfav1.ChallengeExtensions{Scope: mfav1.ChallengeScope_CHALLENGE_SCOPE_USER_SESSION},
+	)
+	if err != nil {
+		return nil, trace.AccessDenied("failed to validate Webauthn response: %v", err)
```

Comment:
FYI, I tend to avoid prefixing errors with "failed to" and similar phrases, as it's easy for the chain to become a sequence of "failed to x: failed to y: failed to ...". It isn't as bad with trace, but it's still good advice.

This item from Uber style describes it too:

* https://github.com/uber-go/guide/blob/master/style.md#error-wrapping`